core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala

```diff
@@ -0,0 +1,156 @@
+  def getExecutorBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String]
+
+  def getNodeBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String]
+
+  // Default implementation to remove failure executors from HashMap based on given time period.
+  def expireExecutorsInBlackList(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]): Unit = {
+    val now = new SystemClock().getTimeMillis()
+    failedExecutorMap.retain((executorid, failureStatus) => {
+      (now - failureStatus.updatedTime) < expireTimeInMillisecond
+    })
+  }
+}
+
+/**
+ * The interface for taskId unrelated strategy. In this case, provided taskId will be ignored. The
+ * benefit for taskId unrelated strategy is different taskSets can learn experience from other
+ * taskSet to avoid allocating tasks to problematic executors.
+ */
+trait TaskUnrelatedBlacklistStrategy extends BlacklistStrategy with Logging{
+  def getExecutorBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String] = {
+    if (taskId.isDefined){
+      logWarning("TaskId unrelated strategy is applied, given taskId will be ignored")
+    }
+    getExecutorBlacklist(failedExecutorMap)
+  }
+
+  def getNodeBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String] = {
+    if (taskId.isDefined){
+      logWarning("TaskId unrelated strategy is applied, given taskId will be ignored")
+    }
+    getNodeBlacklist(failedExecutorMap)
+  }
+
+  // define new interface which remove taskId from parameter.
+  def getExecutorBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]): Set[String]
+  def getNodeBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]): Set[String]
+}
+
+/**
+ * This strategy is simply based on given threshold and is taskId unrelated. An executor will be
+ * in blacklist, if it failed more than "maxFailureTaskNumber" times. A node will be in blacklist,
+ * if there are more than "maxBlackExecutorNumber" executors on it in executor blacklist.
+ */
+class SimpleStrategy(
+    maxFailureTaskNumber: Int,
+    maxBlackExecutorNumber: Int,
+    val expireTimeInMillisecond: Long
+  )extends TaskUnrelatedBlacklistStrategy {
+
+  private def getSelectedExecutorMap(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]) = {
+    failedExecutorMap.filter{
+      case (id, failureStatus) => failureStatus.failureTimes > maxFailureTaskNumber
+    }
+  }
+
+  def getExecutorBlacklist(
+      failureExecutors: mutable.HashMap[String, FailureStatus]): Set[String] = {
+    getSelectedExecutorMap(failureExecutors).keys.toSet
+  }
+
+  def getNodeBlacklist(
+      failureExecutors: mutable.HashMap[String, FailureStatus]): Set[String] = {
```

Comment:
ah, I was thinking before that the runtime would be proportional to the number of executors, but its really just dependent on the executors w/ failed tasks.  Still I think it is probably good to minimize the work done here if we can and its not too hard.  For instance, one easy thing is just avoiding querying the sparkConf inside this function ( currently there is `sparkConf.getBoolean("spark.scheduler.blacklist.speculate", false)`).

I see now that it is a little hard to do the caching, given the separation between `BlacklistTracker` and `BlacklistStrategy`, since `BlacklistStrategy` is responsible for updating `failureExecutors`, and there is also a time-expiration as well.  I dont' see how a `isChanged` boolean is enough to manage this, but maybe I'm wrong -- happy to look at the code.  I was thinking you would need a `failedExecutorVersion: Long`, which gets incremented by the tracker if its ever updated, but maybe there is something simpler.

I will think about this some more ...